rspace/src/main/scala/coop/rchain/rspace/pure/StoreInstances.scala

```diff
@@ -0,0 +1,411 @@
+            Option(ctx.dbJoins.get(txn, joinKey))
+              .map(toByteVectors)
+              .map(_.map(fromByteVectors[C]))
+              .map(exSeq => (exSeq, exSeq.indexOf(channels)))
+              .map {
+                case (exSeq, idx) =>
+                  if (idx >= 0) {
+                    val channelsHash = hashChannelsInner(channels)
+                    if (ctx.dbWaitingContinuations.get(txn, channelsHash) == null) {
+                      val resSeq = util.dropIndex(exSeq, idx)
+                      if (resSeq.nonEmpty) {
+                        ctx.dbJoins.put(txn, joinKey, toByteBuffer(resSeq.map(toByteVectorSeq(_))))
+                      } else {
+                        ctx.dbJoins.delete(txn, joinKey)
+                        collectGarbage(txn, joinKey, joinsCollected = true)
+                      }
+                    }
+                  } else {
+                    throw new IllegalArgumentException(
+                      s"removeJoin: $channels is not a member of $exSeq")
+                  }
+              }
+              .getOrElse(())
+        })
+
+      override def removeAllJoins(txn: T, channel: C): ReaderT[F, LMDBContext, Unit] =
+        ReaderT(ctx =>
+          capture {
+            val joinKey = hashChannelsInner(Seq(channel))
+            ctx.dbJoins.delete(txn, joinKey)
+            collectGarbage(txn, joinKey)
+        })
+
+      override def putJoin(txn: T, channel: C, channels: Seq[C]): ReaderT[F, LMDBContext, Unit] =
+        ReaderT(ctx =>
+          capture {
+            val joinKey = hashChannelsInner(Seq(channel))
+            val oldJoinsBv =
+              Option(ctx.dbJoins.get(txn, joinKey))
+                .map(toByteVectors)
+                .getOrElse(Seq.empty[Seq[ByteVector]])
+
+            val newJoin = toByteVectorSeq(channels)
+            if (!oldJoinsBv.contains(newJoin)) {
+              ctx.dbJoins.put(txn, joinKey, toByteBuffer(newJoin +: oldJoinsBv))
+            }
+        })
+
+      override def getJoin(txn: T, channel: C): ReaderT[F, LMDBContext, Seq[Seq[C]]] =
+        ReaderT(ctx =>
+          capture {
+            val joinKey = hashChannelsInner(Seq(channel))
+            Option(ctx.dbJoins.get(txn, joinKey))
+              .map(toByteVectors)
+              .map(_.map(fromByteVectors[C]))
+              .getOrElse(Seq.empty[Seq[C]])
+        })
+
+      override def removeAll(txn: T, channels: Seq[C]): ReaderT[F, LMDBContext, Unit] =
+        hashChannels(channels)
+          .flatMap(channelsHash => {
+            readWaitingContinuationByteses(txn, channelsHash)
+              .map(_ => writeWaitingContinuationByteses(txn, channelsHash, Seq.empty))
+
+            readDatumByteses(txn, channelsHash)
+              .map(_ => writeDatumByteses(txn, channelsHash, Seq.empty))
+          })
+          .map(_ => channels.foreach(c => removeJoin(txn, c, channels)))
+
+      override def toMap: ReaderT[F, LMDBContext, Map[Seq[C], Row[P, A, K]]] = {
```

Comment:
How about:

```scala
      def toMap: ReaderT[F, LMDBContext, Map[Seq[C], Row[P, A, K]]] = {
        def getIterator(txn: Txn[ByteBuffer]): ReaderT[F, LMDBContext, CursorIterator[ByteBuffer]] =
          ReaderT { ctx =>
            capture {
              val keyRange: KeyRange[ByteBuffer] = KeyRange.all()
              ctx.dbKeys.iterate(txn, keyRange)
            }
          }
        def go(txn: Txn[ByteBuffer]): ReaderT[F, LMDBContext, List[(Seq[C], Row[P, A, K])]] =
          withTxn(txn) { txn =>
            getIterator(txn).flatMap { iterator =>
              withResource(iterator) { (it: CursorIterator[ByteBuffer]) =>
                it.asScala.toList.traverse[ReaderT[F, LMDBContext, ?], (Seq[C], Row[P, A, K])] {
                  (ci: CursorIterator.KeyVal[ByteBuffer]) =>
                    for {
                      channels             <- getChannels(txn, ci.key())
                      data                 <- getData(txn, channels)
                      waitingContinuations <- getWaitingContinuation(txn, channels)
                    } yield channels -> Row(data, waitingContinuations)
                }
              }
            }
          }
        for {
          txn <- createTxnRead()
          res <- go(txn)
        } yield res.toMap
      }
```

The type annotation on `traverse` isn't strictly necessary to get it to compile, but it does help clarify things (and it helps out IntelliJ!)

Truth be told, I think we might have to start moving away from the `withTxn`/`withResource` pattern and start using `bracket`.  @rabbitonweb - thoughts?  Do we have a notion of bracketing for `Capture`?

Reply:
I'm scare of iterator.toList and tried to avoid it in my solutions, expecting that we will face a problem with zillions of records later (that will be solved by adding a filter argument). Is there an easy way to avoid toList or it's not a problem at all?

Reply:
I hear you.  I think it's fine for getting this PR over the fence, but we should consider other options for lazily enumerating store contents (perhaps using Streams).  Effectful Iterators don't really work well in a pure functional context to begin with...

Reply:
Is the `foldLeft` any better in this regard?  I wouldn't think it would be...